substrate/client/rpc/src/utils.rs

```diff
@@ -0,0 +1,245 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! JSON-RPC helpers.
+use crate::SubscriptionTaskExecutor;
+use futures::{
+	future::{self, Either, Fuse, FusedFuture},
+	Future, FutureExt, Stream, StreamExt,
+};
+use jsonrpsee::{PendingSubscriptionSink, SubscriptionMessage, SubscriptionSink};
+use sp_runtime::Serialize;
+use std::collections::VecDeque;
+
+const DEFAULT_BUF_SIZE: usize = 16;
+
+/// A simple bounded VecDeque.
+struct BoundedVecDeque<T> {
+	inner: VecDeque<T>,
+	max_cap: usize,
+}
+
+impl<T> BoundedVecDeque<T> {
+	/// Create a new bounded VecDeque.
+	fn new() -> Self {
+		Self { inner: VecDeque::with_capacity(DEFAULT_BUF_SIZE), max_cap: DEFAULT_BUF_SIZE }
+	}
+
+	fn push_back(&mut self, item: T) -> Result<(), ()> {
+		if self.inner.len() >= self.max_cap {
+			Err(())
+		} else {
+			self.inner.push_back(item);
+			Ok(())
+		}
+	}
+
+	fn pop_front(&mut self) -> Option<T> {
+		self.inner.pop_front()
+	}
+}
+
+/// Feed items to the subscription from the underlying stream.
+///
+/// It's possible configure how many items from the stream
+/// are allowed to be kept in memory until the subscription is dropped.
+///
+/// This is needed because the underlying streams in substrate are
+/// unbounded.
+pub async fn pipe_from_stream<S, T>(pending: PendingSubscriptionSink, mut stream: S)
+where
+	S: Stream<Item = T> + Unpin + Send + 'static,
+	T: Serialize + Send + 'static,
+{
+	let mut buf = BoundedVecDeque::new();
+	let accept_fut = pending.accept();
+
+	futures::pin_mut!(accept_fut);
+
+	// Poll the stream while waiting for the subscription to be accepted
+	//
+	// If the `max_cap` is exceeded then the subscription is dropped.
+	let sink = loop {
+		match future::select(accept_fut, stream.next()).await {
+			Either::Left((Ok(sink), _)) => break sink,
+			Either::Right((Some(msg), f)) => {
+				if buf.push_back(msg).is_err() {
+					log::warn!(target: "rpc", "Subscription::accept failed buffer limit={} exceed; dropping subscription", buf.max_cap);
+					return
+				}
+				accept_fut = f;
+			},
+			// The connection was closed or the stream was closed.
+			_ => return,
+		}
+	};
+
+	inner_pipe_from_stream(sink, stream, buf).await
+}
+
+async fn inner_pipe_from_stream<S, T>(
+	sink: SubscriptionSink,
+	mut stream: S,
+	mut buf: BoundedVecDeque<T>,
+) where
+	S: Stream<Item = T> + Unpin + Send + 'static,
+	T: Serialize + Send + 'static,
+{
+	let mut next_fut = Box::pin(Fuse::terminated());
+	let mut next_item = stream.next();
+	let closed = sink.closed();
+
+	futures::pin_mut!(closed);
+
+	loop {
+		if next_fut.is_terminated() {
+			if let Some(v) = buf.pop_front() {
+				let val = to_sub_message(&sink, &v);
+				next_fut.set(async { sink.send(val).await }.fuse());
+			}
+		}
+
+		match future::select(closed, future::select(next_fut, next_item)).await {
```

Comment:
Nit: Probably just a matter of preference, but one of the tokio or futures select macros might make this a bit easier to read since you don't need to nest then.

Reply:
Yepp, in general agree but tokio::select is easier to read but also a footgun if the future is not cancel-safe.

That's not the case here and I implemented a slightly weird borrow workaround to be able to not drop `next_fut` if another select branch is taken i.e, if pending send operation is pushed to `next_fut` we want not to drop it :)